### PR TITLE
feat: rbac client side code

### PR DESCRIPTION
**Pull request by zlwaterfield**
## Changes

See issue: https://github.com/PostHog/posthog/issues/24512

This PR is adding the main UI for RBAC behind a feature flag. This is all client side and only available for those with the feature flag. This is just the first PR, more to come. This should not affect any existing users.

Added
- organization roles settings
- project access control settings
- resource access control in side panel - for insight, dashboards, feature flags and notebooks
- links to access control from existing dashboard sharing and feature flags

See slack #project-rbac for next steps.

Role management:
Screenshot 2024-12-10 at 11 44 42 AM

Project access:
Screenshot 2024-12-10 at 11 44 38 AM

Resource access control:
Screenshot 2024-12-10 at 11 44 32 AM

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?

It doesn't have an impact.

## How did you test this code?

Manually tested
- without feature flag (free, paid, teams and enterprise)
- with feature flag (free, paid, teams and enterprise)
- tested pay gates as well

### CODE REVIEW

File: frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+                    label: role.name,
+                }))}
+            />
+
+            <LemonTable columns={columns} dataSource={accessControlRoles} loading={accessControlsLoading} />
+        </div>
+    )
+}
+
+// function LevelComponent(member: FusedTeamMemberType): JSX.Element | null {
```
Comment (benjackwhite) [inline frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx:275]: Probs wants to be removed

File: frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+        {
+            title: 'Level',
+            key: 'level',
+            width: 0,
+            render: (_, { access_level, role }) => {
+                return (
+                    <div className="my-1">
+                        <SimplLevelComponent
+                            size="small"
+                            level={access_level}
```
Comment (benjackwhite) [inline frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx:233]: One thing missing here potentially is a notice to indicate if the default access level is higher than the selected level (making it clear that the default level of "editor" overrides "viewer" for example

File: frontend/src/scenes/notebooks/Notebook/notebookLogic.ts [hunk 3]
```diff
@@ -518,6 +528,11 @@ export const notebookLogic = kea<notebookLogicType>([
             )
         },
         setLocalContent: async ({ updateEditor, jsonContent }, breakpoint) => {
+            if (values.notebook?.user_access_level !== 'editor') {
```
Comment (benjackwhite) [inline frontend/src/scenes/notebooks/Notebook/notebookLogic.ts:531]: Edge case - If user A has the notebook open and then user B swaps the permissions to view only user A can get stuck in a state where the local content is there but the API requests failed... We probs need an alert if the API response indicates no permissions and then some sort of refresh.

File: frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+                        <b>You don't have permission to edit access controls for {suffix}.</b>
+                        <br />
+                        You must be the creator of it, a Project Admin, or an Organization Admin.
+                    </LemonBanner>
+                ) : null}
+                <h3>Default access to {suffix}</h3>
+                <AccessControlObjectDefaults />
+
+                <h3>Members</h3>
+                <PayGateMini feature={AvailableFeature.PROJECT_BASED_PERMISSIONING}>
```
Comment (benjackwhite) [inline frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx:52]: It's a pretty nasty loading state here... There is a good few seconds delay due to billing api being so slow and you just see a blank... Could be follow up but I think `PayGateMini` needs a "loading" prop so you can render some skeletons

File: frontend/src/lib/components/RestrictedArea.tsx [hunk 5]
```diff
@@ -37,6 +40,7 @@ export function useRestrictedArea({ scope, minimumAccessLevel }: UseRestrictedAr
             if (!isAuthenticatedTeam(currentTeam)) {
                 return 'Loading current project…'
             }
+            // TODO: Check the user_access_level
```
Comment (benjackwhite) [inline frontend/src/lib/components/RestrictedArea.tsx:43]: Just thinking these TODOs can get forgotten - is in a megaissue somewhere?

File: frontend/src/scenes/actions/actionLogic.ts [hunk 6]
```diff
@@ -106,17 +112,19 @@ export const actionLogic = kea<actionLogicType>([
...
-            (action): ActivityFilters | null => {
-                return action?.id
-                    ? {
-                          scope: ActivityScope.ACTION,
-                          item_id: String(action.id),
-                      }
-                    : null
-            },
-        ],
+        // [SIDE_PANEL_CONTEXT_KEY]: [
```
Comment (benjackwhite) [inline frontend/src/scenes/actions/actionLogic.ts:115]: This probably wants to be added back in right? At least the activity scope if not the access controls

File: frontend/src/scenes/dashboard/DashboardCollaborators.tsx [hunk 7]
```diff
@@ -36,73 +40,96 @@ export function DashboardCollaboration({ dashboardId }: { dashboardId: Dashboard
...
+    if (!dashboard) {
+        return null
+    }
+
+    if (newAccessControl) {
+        return (
+            <div>
+                <h3>Access control</h3>
+                <LemonBanner type="info" className="mb-4">
+                    Permissions have moved! We're rolling out our new access control system. Click below to open it.
```
Comment (benjackwhite) [inline frontend/src/scenes/dashboard/DashboardCollaborators.tsx:57]: This is a nice system

Comment (benjackwhite) [reply frontend/src/scenes/dashboard/DashboardCollaborators.tsx:57]: Although we _may_ need to consider still showing the old UI if there are some configurations set. Depends on whether we will go for an all-in-one migration or not

File: frontend/src/scenes/persons/personsLogic.tsx [hunk 8]
```diff
@@ -256,16 +256,16 @@ export const personsLogic = kea<personsLogicType>([
...
-            (s) => [s.person],
-            (person): ActivityFilters => {
-                return {
-                    scope: ActivityScope.PERSON,
-                    // TODO: Is this correct? It doesn't seem to work...
-                    item_id: person?.id ? `${person?.id}` : undefined,
-                }
-            },
-        ],
+        // [SIDE_PANEL_CONTEXT_KEY]: [
```
Comment (benjackwhite) [inline frontend/src/scenes/persons/personsLogic.tsx:259]: Same here - lets keep the activity scope

File: frontend/src/scenes/pipeline/pipelineLogic.tsx [hunk 9]
```diff
@@ -44,14 +49,14 @@ export const pipelineLogic = kea<pipelineLogicType>([
...
 
-        activityFilters: [
-            () => [],
-            (): ActivityFilters | null => {
-                return {
-                    scope: ActivityScope.PLUGIN,
-                }
-            },
-        ],
+        // [SIDE_PANEL_CONTEXT_KEY]: [
```
Comment (benjackwhite) [inline frontend/src/scenes/pipeline/pipelineLogic.tsx:52]: Same?

File: frontend/src/scenes/pipeline/pipelineNodeLogic.tsx [hunk 10]
```diff
@@ -78,17 +84,19 @@ export const pipelineNodeLogic = kea<pipelineNodeLogicType>([
...
-            (node): ActivityFilters | null => {
-                return node.backend === PipelineBackend.Plugin
-                    ? {
-                          scope: ActivityScope.PLUGIN,
-                          item_id: `${node.id}`,
-                      }
-                    : null
-            },
-        ],
+        // [SIDE_PANEL_CONTEXT_KEY]: [
```
Comment (benjackwhite) [inline frontend/src/scenes/pipeline/pipelineNodeLogic.tsx:87]: You get my point...

File: frontend/src/scenes/settings/organization/Members.tsx [hunk 11]
```diff
@@ -167,7 +169,7 @@ export function Members(): JSX.Element | null {
             title: 'Name',
             key: 'user_name',
             render: (_, member) =>
-                member.user.uuid == user.uuid ? `${fullName(member.user)} (me)` : fullName(member.user),
+                member.user.uuid == user.uuid ? `${fullName(member.user)} (you)` : fullName(member.user),
```
Comment (benjackwhite) [inline frontend/src/scenes/settings/organization/Members.tsx:172]: "You" makes so much more sense

**Review by benjackwhite — APPROVED** [review]
I love this PR so much. No idea if this is because I'm seeing a lot of my own code and having an incredibly vain moment or if it's because your modifications / rewrite is so fantastic. I'm gonna call this a fine cocktail of the both of us called the Ben-Zachary 😍 

All of my comments outside the activity scope stuff can be follow up rather than blocking 🙌

File: frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx (hunk 4 above)
Comment (zlwaterfield) [reply frontend/src/layout/navigation-3000/sidepanel/panels/access_control/AccessControlObject.tsx:52]: Yes this is really needed. Thanks for the ping.

File: frontend/src/scenes/dashboard/DashboardCollaborators.tsx (hunk 7 above)
Comment (zlwaterfield) [reply frontend/src/scenes/dashboard/DashboardCollaborators.tsx:57]: This only renders when the flag is on so my idea was we migrate users and then turn on the flag but that might be too simplistic so I'll keep that in mind.

File: frontend/src/scenes/actions/actionLogic.ts (hunk 6 above)
Comment (zlwaterfield) [reply frontend/src/scenes/actions/actionLogic.ts:115]: I was trying to limit the across of access control for this first iteration. But yes you're correct so will look at making sure the activity scope is still set.

**Pull request merged**